### ISSUE TITLE
High memory usage after the sharding upgrade

### ISSUE THREAD

**Issue by bowenwang1996** [near/nearcore#5072]
**Describe the bug**
We observed on testnet that some rpc nodes run out of memory in less than a day https://nearprotocol.pagerduty.com/incidents/Q3WH1JSK2GGW8B and it has happened twice already.

**To Reproduce**
Run a rpc node.

**Expected behavior**
The node should not crash due to oom

**Screenshots**
dmesg output
```
[3109213.237239] [  pid  ]   uid  tgid total_vm      rss pgtables_bytes swapents oom_score_adj name
[3109213.237241] [1948304]  1000 1948304      654      136    45056        0             0 sh
[3109213.237245] [1948305]  1000 1948305  4440455  3654049 30035968        0             0 neard
[3109213.237250] [1948306]  1000 1948306     2073      557    49152        0             0 sed
[3109213.237253] oom-kill:constraint=CONSTRAINT_MEMCG,nodemask=(null),cpuset=/,mems_allowed=0,oom_memcg=/system.slice/neard.service,task_memcg=/system.slice/neard.service,task=neard,pid=1948305,uid=1000
[3109213.237460] Memory cgroup out of memory: Killed process 1948305 (neard) total-vm:17761820kB, anon-rss:14616196kB, file-rss:0kB, shmem-rss:0kB, UID:1000 pgtables:29332kB oom_score_adj:0
```

**Version (please complete the following information):**
- nearcore https://github.com/near/nearcore/releases/tag/1.22.0-rc.2

**Comment by matklad** [comment]
Prepared fixes:

* https://github.com/near/nearcore/pull/5114 (master)
* https://github.com/near/nearcore/pull/5115 (1.22.0)

**Comment by nikurt** [comment]
Enabling jemalloc improves memory usage a lot.
A node used to OOM every 5 hours, but with jemalloc it OOMs every 20-40 hours.
![image](https://user-images.githubusercontent.com/86772482/140035777-b2b60ca0-3c8d-47f8-a89e-e7c1cd8c4a77.png)

**Comment by bowenwang1996** [comment]
Fixed in #5114
